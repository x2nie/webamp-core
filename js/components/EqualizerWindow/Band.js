import React from "react";
import { connect } from "react-redux";
import Slider from "rc-slider/lib/Slider";
import { SET_BAND_FOCUS, UNSET_FOCUS } from "../../actionTypes";

const MAX_VALUE = 100;
// Given a value between 1-100, return the sprite number (0-27)
export const spriteNumber = value => {
  const percent = value / 100;
  return Math.round(percent * 27);
};

// Given a sprite number, return the x,y
export const spriteOffsets = number => {
  const x = number % 14;
  const y = Math.floor(number / 14);
  return { x, y };
};

const Handle = () => <div className="rc-slider-handle" />;

const Band = ({
  value,
  backgroundPosition,
  id,
  onChange,
  handleMouseDown,
  handleMouseUp
}) => (
  <div id={id} className="band" style={{ backgroundPosition }}>
    <Slider
      type="range"
      min={0}
      max={MAX_VALUE}
      step={1}
      value={MAX_VALUE - value}
      vertical
      onChange={onChange}
      onBeforeChange={handleMouseDown}
      onAfterChange={handleMouseUp}
      handle={Handle}
    />
  </div>
);

const mapStateToProps = (state, ownProps) => {
  const value = state.equalizer.sliders[ownProps.band];
  const { x, y } = spriteOffsets(spriteNumber(value));
  const xOffset = x * 15; // Each sprite is 15px wide
  const yOffset = y * 65; // Each sprite is 15px tall
  const backgroundPosition = `-${xOffset}px -${yOffset}px`;
  return {
    id: ownProps.id,
    value,
    backgroundPosition
  };
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  handleMouseDown: () =>
    dispatch({ type: SET_BAND_FOCUS, input: "eq", bandFocused: ownProps.band }),
  handleMouseUp: () => dispatch({ type: UNSET_FOCUS })
});

export default connect(mapStateToProps, mapDispatchToProps)(Band);
