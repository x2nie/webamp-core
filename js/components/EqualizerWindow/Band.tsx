import React, { useMemo } from "react";
import Slider from "rc-slider";
import { Slider as SliderType } from "../../types";
import { useTypedSelector, useActionCreator } from "../../hooks";
import * as Selectors from "../../selectors";
import * as Actions from "../../actionCreators";

interface Props {
  id: string;
  band: SliderType;
  onChange(value: number): void;
}

const MAX_VALUE = 100;
// Given a value between 1-100, return the sprite number (0-27)
export const spriteNumber = (value: number): number => {
  const percent = value / 100;
  return Math.round(percent * 27);
};

// Given a sprite number, return the x,y
export const spriteOffsets = (number: number): { x: number; y: number } => {
  const x = number % 14;
  const y = Math.floor(number / 14);
  return { x, y };
};

const Handle = () => <div className="rc-slider-handle" />;

export default function Band({ id, onChange, band }: Props) {
  const sliders = useTypedSelector(Selectors.getSliders);
  const value = sliders[band];
  const backgroundPosition = useMemo(() => {
    const { x, y } = spriteOffsets(spriteNumber(value));
    const xOffset = x * 15; // Each sprite is 15px wide
    const yOffset = y * 65; // Each sprite is 15px tall
    return `-${xOffset}px -${yOffset}px`;
  }, [value]);
  const focusBand = useActionCreator(Actions.focusBand);
  const usetFocus = useActionCreator(Actions.unsetFocus);

  const handleMouseDown = () => focusBand(band);

  return (
    <div id={id} className="band" style={{ backgroundPosition }}>
      <Slider
        min={0}
        max={MAX_VALUE}
        step={1}
        value={MAX_VALUE - value}
        vertical
        onChange={onChange}
        onBeforeChange={handleMouseDown}
        onAfterChange={usetFocus}
        handle={Handle}
      />
    </div>
  );
}
