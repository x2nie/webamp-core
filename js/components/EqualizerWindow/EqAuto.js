import React from "react";
import { connect } from "react-redux";
import classnames from "classnames";

import { SET_EQ_AUTO } from "../../actionTypes";

const EqAuto = props => {
  const className = classnames({
    selected: props.auto
  });
  return <div id="auto" className={className} onClick={props.toggleAuto} />;
};

const toggleAuto = () => (dispatch, getState) => {
  dispatch({ type: SET_EQ_AUTO, value: !getState().equalizer.auto });
};

const mapDispatchToProps = { toggleAuto };
export default connect(
  state => ({ toggleAuto: state.equalizer.toggleAuto }),
  mapDispatchToProps
)(EqAuto);
