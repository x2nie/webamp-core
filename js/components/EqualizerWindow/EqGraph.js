import React from "react";
import { connect } from "react-redux";
import { getCurvePoints } from "cardinal-spline-js";
import { rebound } from "../../utils";

import { BANDS } from "../../constants";

export const roundToEven = value => 2 * Math.round(value / 2);

// The line is two pixels wide, and sits "above" the offset specified. So we use 30 not 32 as our max.
const reboundTo32 = rebound(1, 100, 0, 30);

export const getY = value => roundToEven(reboundTo32(value));

class EqGraph extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount() {
    this.canvasCtx = this.canvas.getContext("2d");
    this.canvasCtx.imageSmoothingEnabled = false;
    this.width = this.canvas.width * 1; // Cast to int
    this.height = this.canvas.height * 1; // Cast to int

    if (this.props.lineColorsImage) {
      this.createColorPattern(this.props.lineColorsImage);
    }
    if (this.props.preampLineUrl) {
      this.createPreampLineImage(this.props.preampLineUrl);
    }
  }

  componentDidUpdate() {
    this.canvasCtx.clearRect(0, 0, this.width, this.height);
    this.drawPreampLine();
    this.drawEqLine(); // This should paint on top of the preamp line
  }

  shouldComponentUpdate(nextProps) {
    if (this.props.lineColorsImage !== nextProps.lineColorsImage) {
      this.createColorPattern(nextProps.lineColorsImage);
    }
    if (this.props.preampLineUrl !== nextProps.preampLineUrl) {
      this.createPreampLineImage(nextProps.preampLineUrl);
    }
    return true;
  }

  createPreampLineImage(preampLineUrl) {
    const preampLineImg = new Image();
    preampLineImg.onload = () => {
      this.setState({ preampLineImg });
    };
    preampLineImg.src = preampLineUrl;
  }

  createColorPattern(lineColorsImage) {
    const bgImage = new Image();
    bgImage.onload = () => {
      const colorsCanvas = document.createElement("canvas");
      const colorsCtx = colorsCanvas.getContext("2d");
      colorsCanvas.width = bgImage.width * 2;
      colorsCanvas.height = bgImage.height * 2;
      colorsCtx.drawImage(
        bgImage,
        0,
        0,
        colorsCanvas.width,
        colorsCanvas.height
      );
      this.setState({
        colorPattern: this.canvasCtx.createPattern(colorsCanvas, "repeat-x")
      });
    };
    bgImage.src = lineColorsImage;
  }

  drawEqLine() {
    if (!this.state.colorPattern) {
      // The skin has not finished loading yet
      return;
    }
    const { props } = this;
    const amplitudes = BANDS.map(band => props[band]);

    this.canvasCtx.strokeStyle = this.state.colorPattern;
    this.canvasCtx.lineWidth = 2;
    this.canvasCtx.beginPath();
    const paddingLeft = 4;

    const points = amplitudes.reduce(
      (prev, value, i) => prev.concat(paddingLeft + i * 16, getY(100 - value)),
      []
    );

    // Spline between points in order to create nice curves
    const tension = 0.5;
    const resolution = 4; // Points in each segment
    const smoothPoints = getCurvePoints(points, tension, resolution);
    for (let i = 0; i < smoothPoints.length; i += 2) {
      this.canvasCtx.lineTo(smoothPoints[i], smoothPoints[i + 1]);
    }

    this.canvasCtx.stroke();
  }

  drawPreampLine() {
    const { preampLineImg } = this.state;
    if (!preampLineImg) {
      // The skin has not finished loading yet
      return;
    }
    const preampValue = getY(this.props.preamp);
    this.canvasCtx.drawImage(
      this.state.preampLineImg,
      0,
      preampValue,
      preampLineImg.width * 2,
      preampLineImg.height * 2
    );
  }

  render() {
    return (
      <canvas
        id="eqGraph"
        ref={node => (this.canvas = node)}
        width="152"
        height="32"
      />
    );
  }
}

export default connect(state => ({
  ...state.equalizer.sliders,
  lineColorsImage: state.display.skinImages.EQ_GRAPH_LINE_COLORS,
  preampLineUrl: state.display.skinImages.EQ_PREAMP_LINE
}))(EqGraph);
