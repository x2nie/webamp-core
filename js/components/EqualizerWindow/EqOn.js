import React from "react";
import { connect } from "react-redux";
import classnames from "classnames";

import { toggleEq } from "../../actionCreators";

const EqOn = props => {
  const className = classnames({
    selected: props.on
  });
  return <div id="on" className={className} onClick={props.toggleEq} />;
};

const mapStateToProps = state => ({ on: state.equalizer.on });

export default connect(mapStateToProps, { toggleEq })(EqOn);
