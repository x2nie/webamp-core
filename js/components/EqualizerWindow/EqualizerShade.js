import React from "react";
import { connect } from "react-redux";
import Volume from "../Volume";
import Balance from "../Balance";
import { segment } from "../../utils";

import {
  closeEqualizerWindow,
  toggleEqualizerShadeMode
} from "../../actionCreators";

const EqualizerShade = props => {
  const { volume, balance } = props;

  const classes = ["left", "center", "right"];
  const eqVolumeClassName = segment(0, 100, volume, classes);
  const eqBalanceClassName = segment(-100, 100, balance, classes);
  return (
    <div className="draggable">
      <div id="equalizer-shade" onClick={props.toggleEqualizerShadeMode} />
      <div id="equalizer-close" onClick={props.closeEqualizerWindow} />
      <Volume id="equalizer-volume" className={eqVolumeClassName} />
      <Balance id="equalizer-balance" className={eqBalanceClassName} />
    </div>
  );
};

const mapDispatchToProps = {
  closeEqualizerWindow,
  toggleEqualizerShadeMode
};

const mapStateToProps = state => ({
  volume: state.media.volume,
  balance: state.media.balance
});

export default connect(mapStateToProps, mapDispatchToProps)(EqualizerShade);
