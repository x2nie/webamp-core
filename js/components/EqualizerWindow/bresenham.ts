// Adapted from https://github.com/nquicenob/bresenham-line by Nicolas Quiceno
interface Point {
  x: number;
  y: number;
}

type Sign = 1 | -1;
function getSing(num: number): Sign {
  return num > 0 ? 1 : -1;
}

function getInitValues(startPoint: Point, finalPoint: Point) {
  const abs = Math.abs;
  const diffx = finalPoint.x - startPoint.x;
  const diffy = finalPoint.y - startPoint.y;

  return {
    absDiff: {
      x: abs(diffx),
      y: abs(diffy)
    },
    sign: {
      x: getSing(diffx),
      y: getSing(diffy)
    }
  };
}

function getBreakFn(sign: Sign): (current: number, final: number) => boolean {
  return sign < 0
    ? (current, final) => current >= final
    : (current, final) => current <= final;
}

function calcMainCoordinates(absDiff: Point): ["x", "y"] | ["y", "x"] {
  return absDiff.x > absDiff.y ? ["x", "y"] : ["y", "x"];
}

export default function line(point: Point, finalPoint: Point) {
  const { absDiff, sign } = getInitValues(point, finalPoint);
  const [mainCoordinate, coordinate] = calcMainCoordinates(absDiff);

  const final = finalPoint[mainCoordinate];

  const mainSign = sign[mainCoordinate];
  const secondSign = sign[coordinate];

  const mainDiff = absDiff[mainCoordinate];
  const secondDiff = absDiff[coordinate];

  const breakFn = getBreakFn(mainSign);

  let mainValue = point[mainCoordinate];
  let secondValue = point[coordinate];

  let eps = 0;

  const points = [];
  for (; breakFn(mainValue, final); mainValue += mainSign) {
    points.push({
      [mainCoordinate]: mainValue,
      [coordinate]: secondValue
    });
    eps += secondDiff;
    if (eps << 1 >= mainDiff) {
      secondValue += secondSign;
      eps -= mainDiff;
    }
  }
  return points;
}
