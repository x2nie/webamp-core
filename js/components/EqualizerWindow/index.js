import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import classnames from "classnames";
import Volume from "../Volume";
import Balance from "../Balance";
import { segment } from "../../utils";

import { BANDS, WINDOWS } from "../../constants";
import {
  setEqBand,
  setPreamp,
  setEqToMax,
  setEqToMid,
  setEqToMin,
  closeEqualizerWindow,
  toggleEqualizerShadeMode
} from "../../actionCreators";

import { SET_FOCUSED_WINDOW } from "../../actionTypes";

import Band from "./Band";
import EqOn from "./EqOn";
import EqAuto from "./EqAuto";
import EqGraph from "./EqGraph";
import PresetsContextMenu from "./PresetsContextMenu";

import "../../../css/equalizer-window.css";

const bandClassName = band => `band-${band}`;

const EqualizerWindow = props => {
  const { doubled, selected, volume, balance, shade } = props;

  const className = classnames({
    selected,
    doubled,
    shade,
    window: true,
    draggable: true
  });

  const classes = ["left", "center", "right"];
  const eqVolumeClassName = segment(0, 100, volume, classes);
  const eqBalanceClassName = segment(-100, 100, balance, classes);
  return (
    <div
      id="equalizer-window"
      className={className}
      onMouseDown={props.focusWindow}
    >
      {props.shade ? (
        <div className="draggable">
          <div id="equalizer-shade" onClick={props.toggleEqualizerShadeMode} />
          <div id="equalizer-close" onClick={props.closeEqualizerWindow} />
          <Volume id="equalizer-volume" className={eqVolumeClassName} />
          <Balance id="equalizer-balance" className={eqBalanceClassName} />
        </div>
      ) : (
        <div>
          <div className="equalizer-top title-bar draggable">
            <div
              id="equalizer-shade"
              onClick={props.toggleEqualizerShadeMode}
            />
            <div id="equalizer-close" onClick={props.closeEqualizerWindow} />
          </div>
          <EqOn />
          <EqAuto />
          <EqGraph />
          <PresetsContextMenu fileInput={props.fileInput} />
          <Band id="preamp" band="preamp" onChange={props.setPreampValue} />
          <div id="plus12db" onClick={props.setEqToMax} />
          <div id="zerodb" onClick={props.setEqToMid} />
          <div id="minus12db" onClick={props.setEqToMin} />
          {BANDS.map(hertz => (
            <Band
              key={hertz}
              id={bandClassName(hertz)}
              band={hertz}
              onChange={props.setHertzValue(hertz)}
            />
          ))}
        </div>
      )}
    </div>
  );
};

EqualizerWindow.propTypes = {
  doubled: PropTypes.bool.isRequired,
  selected: PropTypes.bool.isRequired,
  shade: PropTypes.bool.isRequired
};

const mapDispatchToProps = dispatch => ({
  focusWindow: () =>
    dispatch({ type: SET_FOCUSED_WINDOW, window: WINDOWS.EQUALIZER }),
  setPreampValue: value => dispatch(setPreamp(value)),
  setEqToMin: () => dispatch(setEqToMin()),
  setEqToMid: () => dispatch(setEqToMid()),
  setEqToMax: () => dispatch(setEqToMax()),
  setHertzValue: hertz => value => dispatch(setEqBand(hertz, value)),
  closeEqualizerWindow: () => dispatch(closeEqualizerWindow()),
  toggleEqualizerShadeMode: () => dispatch(toggleEqualizerShadeMode())
});

const mapStateToProps = state => ({
  doubled: state.display.doubled,
  selected: state.windows.focused === WINDOWS.EQUALIZER,
  shade: state.display.equalizerShade,
  volume: state.media.volume,
  balance: state.media.balance
});

export default connect(mapStateToProps, mapDispatchToProps)(EqualizerWindow);
