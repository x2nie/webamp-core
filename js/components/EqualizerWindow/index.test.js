const media = {
  addEventListener: jest.fn(),
  _analyser: null
};

const canvasMockify = require("canvas-mock");

function createNodeMock(element) {
  if (element.type === "canvas") {
    const Canvas = window.document.createElement("canvas");
    canvasMockify(Canvas); // mock canvas functions required by Phaser.js are added
    return Canvas;
  }
  return null;
}

window.requestAnimationFrame = () => {};

import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import getStore from "../../store";

import EqualizerWindow from "./index";

describe("EqualizerWindow", () => {
  let store;
  beforeEach(() => {
    store = getStore(media);
  });

  it("renders to snapshot", () => {
    const options = { createNodeMock };
    const tree = renderer
      .create(
        <Provider store={store}>
          <EqualizerWindow fileInput={null} />
        </Provider>,
        options
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
