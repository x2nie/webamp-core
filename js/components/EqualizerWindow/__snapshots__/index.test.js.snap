// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EqualizerWindow renders to snapshot 1`] = `
<div
  className="window draggable"
  id="equalizer-window"
  onMouseDown={[Function]}
>
  <div>
    <div
      className="equalizer-top title-bar draggable"
    >
      <div
        id="equalizer-shade"
        onClick={[Function]}
      />
      <div
        id="equalizer-close"
        onClick={[Function]}
      />
    </div>
    <div
      className="selected"
      id="on"
      onClick={[Function]}
    />
    <div
      className=""
      id="auto"
      onClick={[Function]}
    />
    <canvas
      height="32"
      id="eqGraph"
      width="152"
    />
    <div
      id="presets-context"
    >
      <div
        className="handle"
        onClick={[Function]}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <div
          id="presets"
        />
      </div>
    </div>
    <div
      className="band"
      id="preamp"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      id="plus12db"
      onClick={[Function]}
    />
    <div
      id="zerodb"
      onClick={[Function]}
    />
    <div
      id="minus12db"
      onClick={[Function]}
    />
    <div
      className="band"
      id="band-60"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-170"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-310"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-600"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-1000"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-3000"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-6000"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-12000"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-14000"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="band"
      id="band-16000"
      style={
        Object {
          "backgroundPosition": "-0px -65px",
        }
      }
    >
      <div
        className="rc-slider rc-slider-vertical"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <div
          className="rc-slider-rail"
          style={Object {}}
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "bottom": "0%",
              "height": "50%",
              "visibility": "visible",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          className="rc-slider-handle"
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
  </div>
</div>
`;
